File: c++/bestTimeToBuyAndSellStock#121/try1Optimal.cpp
// This is the optimal solution
class Solution {
public:
    int maxProfit(vector<int>& prices) 
    {
        int maxProfit = 0;
        if(prices.size() <= 1)
        {
            return 0;
        }
        if(prices.size() == 2)
        {
            maxProfit = prices[1] - prices[0];
            if(maxProfit <= 0)
            {
                return 0;
            }
            return maxProfit;
        }
        for(int left = 0, right = left + 1; right < prices.size(); right++)
        {
            if(prices[right] - prices[left] > maxProfit)
            {
                maxProfit = prices[right] - prices[left];
            }
            if(prices[right] < prices[left])
            {
                left = right; 
            }

        }
        if(maxProfit <= 0)
        {
            return 0;
        }
        return maxProfit;
    }
};

//Time Complexity (O(n))
//Space Complexity (O(1))

File: python/courseScheduleII#210/try1Optimal.py
# This is the optimal solution
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        numDependencies = [0] * numCourses
        dependencyGraph = defaultdict(list)
        ordering = []
        q = deque()
        
        for a, b in prerequisites:
            dependencyGraph[b].append(a)
            numDependencies[a] += 1
        
        for i in range(len(numDependencies)):
            if numDependencies[i] == 0:
                ordering.append(i)
                q.append(i)
        
        while q:
            cur = q.popleft()
            if cur in dependencyGraph:
                for num in dependencyGraph[cur]:
                    numDependencies[num] -= 1
                    if numDependencies[num] == 0:
                        q.append(num)
                        ordering.append(num)
        
        return ordering if len(ordering) == numCourses else []

# Time: O(N) / O(V + E)
# Space: O(N) / O(V + E)
